### PR TITLE
ioSession: Simplify token management

### PR DESCRIPTION
**Pull request by jcvenegas**
proxy: unregisterVM free token
proxy: Remove tokenToSession map from proxy.
client: Remove token from client struct
token: remove tokeninfo struct

Fixes: #66

### CODE REVIEW

**Comment by sboeuf** [comment]
I really need to spend some time on this. Please don't merge before I review it.

**Comment by sboeuf** [comment]
@jcvenegas what is the status of this PR, still WIP ?
Also, is that a release-gating issue ? I want to know in order to understand if we want to spend some time on this for the next week(s).

**Comment by jcvenegas** [comment]
@sboeuf  I remember I've seen some failing tests in semaphore related with `cri-o`. But last semaphore build passed. That is why I marked as `WIP`. It is ready for review.

The issues is marked as `beta gating` and  the proxy is 'leaking'/not garbage collecting iosessions after the VM finish.

**Comment by sboeuf** [comment]
Ok I'll review it

File: proxy.go [hunk 1]
```diff
@@ -70,8 +56,8 @@ type proxy struct {
 	// vms are hashed by their containerID
 	vms map[string]*vm
 
-	// tokenToVM maps I/O token to their per-token info
-	tokenToVM map[Token]*tokenInfo
+	// tokenToSession maps I/O token to their per-token info
```
Comment (jodh-intel) [inline proxy.go]: The comment is now incorrect.

File: vm.go [hunk 2]
```diff
@@ -103,6 +103,8 @@ type ioSession struct {
 	// started. This is used to stop the shim from sending stdin data to a
 	// non-existent process.
 	processStarted chan interface{}
+
+	state tokenState
```
Comment (jodh-intel) [inline vm.go]: It would be good to put a comment on this, for consistency with the other fields.

File: proxy.go [hunk 3]
```diff
@@ -158,34 +140,38 @@ func (proxy *proxy) allocateTokens(vm *vm, numIOStreams int) (*api.IOResponse, e
 	}, nil
 }
 
-func (proxy *proxy) claimToken(token Token) (*tokenInfo, error) {
+func (proxy *proxy) claimToken(token Token) (*ioSession, error) {
```
Comment (jodh-intel) [inline proxy.go]: Although it might not strictly be part of this PR, should this function check for `token == nilToken` as `releaseToken()` does?

**Comment by sboeuf** [comment]
@jcvenegas Okay I am trying to review this PR but I would really appreciate more comments and bigger commit messages to explain the reason why we can do those simplifications. I am not saying this PR is wrong but I want to make sure we're not missing something here. If we can simplify those structures, we should be able to explain it properly.

**Comment by sboeuf** [comment]
And apparently something is wrong because Semaphore and Jenkins are consistently failing on the same CRI-O test (15).

**Comment by amshinde** [comment]
@jcvenegas  What is the status on this PR? Can you address @sboeuf's comment for adding verbose commit messages explaining the reasoning for the simplifications. It will help review the PR.

**Comment by sboeuf** [comment]
@jcvenegas any update/progress on this one ?

**Comment by sboeuf** [comment]
@jcvenegas I think we should close this PR. What do you think ?

**Comment by jcvenegas** [comment]
@yes, closing PR.

**Pull request closed without merging**